File: artifacts/api-server/src/bot/services/nickname.ts
import { type Guild, type GuildMember } from "discord.js";
import { logger } from "../../lib/logger";

const MAX_NICK_LEN = 32;

export function formatNickname(
  baseName: string,
  gen: number,
  position: string,
): string {
  const suffix = ` | ${gen} | ${position}`;
  const allowed = MAX_NICK_LEN - suffix.length;
  if (allowed <= 0) {
    return baseName.slice(0, MAX_NICK_LEN);
  }
  const trimmed = baseName.length > allowed ? baseName.slice(0, allowed).trim() : baseName;
  return `${trimmed}${suffix}`;
}

export function extractBaseName(currentNick: string | null | undefined): string | null {
  if (!currentNick) return null;
  const idx = currentNick.indexOf(" | ");
  if (idx === -1) return currentNick;
  return currentNick.slice(0, idx).trim();
}

export async function fetchMember(
  guild: Guild | null | undefined,
  userId: string,
): Promise<GuildMember | null> {
  if (!guild) return null;
  try {
    return await guild.members.fetch(userId);
  } catch (err) {
    logger.warn({ err, userId }, "Üye getirilemedi");
    return null;
  }
}

export async function applyNickname(
  member: GuildMember,
  baseName: string,
  gen: number,
  position: string,
): Promise<{ ok: boolean; reason?: string; nickname: string }> {
  const nick = formatNickname(baseName, gen, position);
  try {
    await member.setNickname(nick, "Türk Ligi Bot — oyuncu kaydı/güncellemesi");
    return { ok: true, nickname: nick };
  } catch (err) {
    const msg = (err as Error).message ?? "bilinmeyen hata";
    logger.warn(
      { err, userId: member.id, nick },
      "Discord takma adı değiştirilemedi",
    );
    return { ok: false, reason: msg, nickname: nick };
  }
}

File: artifacts/api-server/utils/nickname.py
"""Discord takma ad yönetimi."""
import discord

MAX_NICK_LEN = 32

TURKCE_MEVKI: dict[str, str] = {
    "GK":  "KLC",
    "CB":  "STP",
    "LB":  "SLB",
    "RB":  "SĞB",
    "LWB": "SKB",
    "RWB": "SĞK",
    "CDM": "DOS",
    "CM":  "OS",
    "CAM": "HOS",
    "LM":  "SLO",
    "RM":  "SĞO",
    "LW":  "SAL",
    "RW":  "SAK",
    "ST":  "FRV",
    "CF":  "SAN",
}


def position_tr(position: str) -> str:
    return TURKCE_MEVKI.get(position.upper().strip(), position.upper().strip())


def format_nickname(base_name: str, gen: int, position: str) -> str:
    mevki = position_tr(position)
    suffix = f" ┃ {gen}G ┃ {mevki}"
    allowed = MAX_NICK_LEN - len(suffix)
    if allowed <= 0:
        return base_name[:MAX_NICK_LEN]
    trimmed = base_name[:allowed].rstrip() if len(base_name) > allowed else base_name
    return f"{trimmed}{suffix}"


def extract_base_name(current_nick: str | None) -> str | None:
    if not current_nick:
        return None
    for sep in (" ┃ ", " | "):
        idx = current_nick.find(sep)
        if idx != -1:
            return current_nick[:idx].strip()
    return current_nick


async def apply_nickname(
    member: discord.Member, base_name: str, gen: int, position: str
) -> tuple[bool, str, str | None]:
    """Returns (ok, nickname, error_reason)."""
    nick = format_nickname(base_name, gen, position)
    try:
        await member.edit(nick=nick, reason="Türk Ligi Bot — oyuncu güncellemesi")
        return True, nick, None
    except discord.Forbidden:
        return False, nick, "Yetki yetersiz (ManageNicknames)"
    except Exception as e:
        return False, nick, str(e)


async def sync_player_nickname(guild: discord.Guild | None, player: dict):
    if not guild or not player.get("discord_user_id"):
        return
    try:
        member = await guild.fetch_member(int(player["discord_user_id"]))
    except Exception:
        return
    base = extract_base_name(member.display_name) or player.get("name") or member.name
    await apply_nickname(member, base, player["rating"], player["position"])
